docs/debugging.rst

```diff
@@ -0,0 +1,249 @@
+----------------------------
+
+We may use ``assert`` statement in Taichi-scope. When assertion condition failed, a
+``RuntimeError`` will be raised to indicate error.
+
+To make ``assert`` work, first make sure you are using the **CPU backend**.
+For performance reason, ``assert`` only works when ``debug`` mode is on, For example:
+
+.. code-block:: python
+
+    ti.init(arch=ti.cpu, debug=True)
+
+    x = ti.var(ti.f32, 128)
+
+    @ti.kernel
+    def do_sqrt_all():
+        for i in x:
+            assert x[i] >= 0
+            x[i] = ti.sqrt(x)
+
+
+When your debugging work is done, simply set ``debug=False``. Now ``assert`` will be ignored
+therefore no runtime overhead, making your program easy to debug without losing performance.
+
+
+Compile-time ``ti.static_assert``
+---------------------------------
+
+Like ``ti.static_print``, we also provide a static version of ``assert``:
+``ti.static_assert``, it can be useful to make assertion on data type / dimention / shape.
+It works no matter whether ``debug=True`` is specified. When assertion fails, it will
+raise ``AssertionError`` as a Python-scope ``assert`` does.
+
+For example:
+
+.. code-block:: python
+
+    @ti.func
+    def is_odd(x: ti.template()):
+        ti.static_assert(x.data_type() == ti.i32, "is_odd() is only supported for i32")
+        return x % 2 == 1
+
+
+Tips for debugging
+------------------
+
+Debugging a Taichi program can be hard even with the builtin tools above.
+Taichi developers are currently devoting themselves in improving error messages and warnings
+to help user find potential BUGs in their programs.
+
+Here we collected some common BUGs that one might encounter with a Taichi program:
+
+Static typing system
+++++++++++++++++++++
+
+Taichi pretends that it's a dynamical-typed language like Python, but it's actually a
+statically-typed language which will be translated into high performance CPU/GPU instructions.
+
+So the code behavior in Taichi-scope is actually very different from Python-scope!
+
+Type of a variable is simply **determined at its first initialization and never changes later**.
+
+Although static-type provides better performance and simplicity, but may leads to BUGs if
+users not distinguished Taichi-scope from Python-scope, e.g.:
+
+.. code-block:: python
+
+    @ti.kernel
+    def buggy():
+        ret = 0  # 0 is a integer, so `ret` is typed as int32
+        for i in range(3):
+            ret += 0.1 * i  # i32 += f32, the result is still stored in int32!
+        print(ret)  # will shows 0
+
+    buggy()
+
+The code above shows a common BUG due to the limitation of the static-type system.
+The Taichi compiler should shows a warning like:
+
+.. code-block:: none
+
+    [W 06/27/20 21:43:51.853] [type_check.cpp:visit@66] [$19] Atomic add (float32 to int32) may lose precision.
+
+This means that it cannot store a ``float32`` result to ``int32``.
+The solution is to type ``ret`` as ``float32`` at the first place:
```

Comment:
```suggestion
The solution is to initialize ``ret`` as a float-point value:
```

Note that it can be either `f32` or `f64`, depending on the default float-point precision.